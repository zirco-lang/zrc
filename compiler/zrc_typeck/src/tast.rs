//! Type-checked AST for the Zirco programming language
//!
//! This is similar to the [standard AST](zrc_parser::ast), but with type
//! information attached to it.
//!
//! Note that we only attach source locations to code that would exist after code generation,
//! meaning inlined types do not get a span.

pub mod expr;
pub mod stmt;
pub mod ty;
