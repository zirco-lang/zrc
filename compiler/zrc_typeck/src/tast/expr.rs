//! Expression representation for the Zirco [TAST](super)

use std::{fmt::Display, string::ToString};

pub use zrc_parser::{
    ast::expr::{Arithmetic, BinaryBitwise, Comparison, Equality, Logical},
    lexer::StringTok,
};

/// An [expression kind](TypedExprKind) with its yielded [result
/// type](super::ty::Type) attached to it.
#[derive(PartialEq, Debug, Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct TypedExpr<'input>(pub super::ty::Type<'input>, pub TypedExprKind<'input>);

/// The left hand side of an assignment.
#[derive(PartialEq, Debug, Clone)]
pub struct Place<'input>(pub super::ty::Type<'input>, pub PlaceKind<'input>);
/// The valid left-hand-side of a [`TypedExprKind::Assignment`].
///
/// Places may be:
/// - A variable or an property access of a place
/// - A dereference or index into any expression yielding a pointer
#[derive(PartialEq, Debug, Clone)]
pub enum PlaceKind<'input> {
    /// `*x`
    Deref(Box<TypedExpr<'input>>),
    /// `x`
    Variable(&'input str),
    /// `x[y]`
    Index(Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// `x.y`
    Dot(Box<Place<'input>>, &'input str),
}

/// The kind of a [`TypedExpr`]
#[derive(PartialEq, Debug, Clone)]
pub enum TypedExprKind<'input> {
    /// `a, b`
    Comma(Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),

    /// `a = b`
    ///
    /// `a += b` is desugared to `a = a + b` by the type checker.
    Assignment(Box<Place<'input>>, Box<TypedExpr<'input>>),

    /// Bitwise operations
    BinaryBitwise(
        BinaryBitwise,
        Box<TypedExpr<'input>>,
        Box<TypedExpr<'input>>,
    ),
    /// Logical operations
    Logical(Logical, Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// Equality checks
    Equality(Equality, Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// Comparisons
    Comparison(Comparison, Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// Arithmetic operations
    Arithmetic(Arithmetic, Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),

    /// `!x`
    UnaryNot(Box<TypedExpr<'input>>),
    /// `~x`
    UnaryBitwiseNot(Box<TypedExpr<'input>>),
    /// `-x`
    UnaryMinus(Box<TypedExpr<'input>>),
    /// `&x`
    UnaryAddressOf(Box<Place<'input>>),
    /// `*x`
    UnaryDereference(Box<TypedExpr<'input>>),

    /// `a[b]`
    Index(Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// `a.b`
    Dot(Box<Place<'input>>, &'input str),
    /// `a(b, c, d, ...)`
    Call(Box<Place<'input>>, Vec<TypedExpr<'input>>),

    /// `a ? b : c`
    Ternary(
        Box<TypedExpr<'input>>,
        Box<TypedExpr<'input>>,
        Box<TypedExpr<'input>>,
    ),

    /// `x as T`
    Cast(Box<TypedExpr<'input>>, super::ty::Type<'input>),

    /// Any numeric literal.
    NumberLiteral(&'input str),
    /// Any string literal.bool
    StringLiteral(Vec<StringTok<'input>>),
    /// Any identifier.
    Identifier(&'input str),
    /// Any boolean literal.
    BooleanLiteral(bool),
}

impl<'input> Display for TypedExprKind<'input> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NumberLiteral(n) => write!(f, "{n}"),
            Self::StringLiteral(str) => write!(
                f,
                "\"{}\"",
                str.iter().map(ToString::to_string).collect::<String>()
            ),
            Self::Identifier(i) => write!(f, "{i}"),
            Self::BooleanLiteral(value) => write!(f, "{value}"),
            Self::Assignment(place, value) => write!(f, "{place} = {value}"),
            Self::Equality(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::Comparison(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::Arithmetic(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::BinaryBitwise(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::Logical(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::Comma(lhs, rhs) => write!(f, "{lhs}, {rhs}"),
            Self::UnaryNot(expr) => write!(f, "!{expr}"),
            Self::UnaryBitwiseNot(expr) => write!(f, "~{expr}"),
            Self::UnaryMinus(expr) => write!(f, "-{expr}"),
            Self::UnaryAddressOf(expr) => write!(f, "&{expr}"),
            Self::UnaryDereference(expr) => write!(f, "*{expr}"),
            Self::Ternary(cond, if_true, if_false) => write!(f, "{cond} ? {if_true} : {if_false}"),
            Self::Index(ptr, idx) => write!(f, "{ptr}[{idx}]"),
            Self::Dot(expr, key) => write!(f, "{expr}.{key}"),
            Self::Cast(expr, ty) => write!(f, "{expr} as {ty}"),
            Self::Call(expr, args) => write!(
                f,
                "{expr}({})",
                args.iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
        }
    }
}
impl<'input> Display for PlaceKind<'input> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Deref(expr) => write!(f, "*{expr}"),
            Self::Variable(ident) => write!(f, "{ident}"),
            Self::Index(ptr, idx) => write!(f, "{ptr}[{idx}]"),
            Self::Dot(expr, key) => write!(f, "{expr}.{key}"),
        }
    }
}
impl<'input> Display for Place<'input> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "({}) as ({})", self.1, self.0)
    }
}

impl<'input> Display for TypedExpr<'input> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(({}) as ({}))", self.1, self.0)
    }
}
