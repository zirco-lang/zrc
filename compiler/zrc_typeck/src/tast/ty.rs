//! Type representation for the Zirco [TAST](super)

use std::{collections::HashMap, fmt::Display};

use crate::typeck::BlockReturnType;

/// The possible Zirco types
#[derive(PartialEq, Debug, Clone)]
pub enum Type {
    // WHENEVER ADDING NEW PRIMITIVES HERE, ADD THEM TO THE TYPE SCOPE IN
    // [`zrc_typeck::typeck::Scope::default`].
    /// `i8`
    I8,
    /// `u8`
    U8,
    /// `i16`
    I16,
    /// `u16`
    U16,
    /// `i32`
    I32,
    /// `u32`
    U32,
    /// `i64`
    I64,
    /// `u64`
    U64,
    /// `void`, only producible by calling a void function (`fn()`)
    Void,
    /// `bool`
    Bool, /* TODO: need an "any Int" type that implicitly casts to all int types but becomes
           * i32 when assigned to a value */
    /// `*T`
    Ptr(Box<Type>),
    /// `fn(A, B) -> T`
    Fn(Vec<Type>, Box<BlockReturnType>),
    /// Struct type literals
    Struct(HashMap<String, Type>),
}

impl Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.clone() {
            Self::I8 => write!(f, "i8"),
            Self::U8 => write!(f, "u8"),
            Self::I16 => write!(f, "i16"),
            Self::U16 => write!(f, "u16"),
            Self::I32 => write!(f, "i32"),
            Self::U32 => write!(f, "u32"),
            Self::I64 => write!(f, "i64"),
            Self::U64 => write!(f, "u64"),
            Self::Bool => write!(f, "bool"),
            Self::Ptr(t) => write!(f, "*({t})"),
            Self::Void => write!(f, "void"),
            Self::Fn(args, brt) => write!(
                f,
                "(fn({}){})",
                args.iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", "),
                match *brt {
                    BlockReturnType::Return(ret) => format!(" -> {ret}"),
                    BlockReturnType::Void => String::new(),
                }
            ),
            Self::Struct(fields) => write!(
                f,
                "(struct {{ {} }})",
                fields
                    .iter()
                    .map(|(k, v)| format!("{k}: {v}"))
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
        }
    }
}

impl Type {
    /// Returns `true` if this is an integer type like [`Type::I8`].
    #[must_use]
    pub const fn is_integer(&self) -> bool {
        use Type::{I16, I32, I64, I8, U16, U32, U64, U8};
        matches!(self, I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64)
    }

    /// Returns `true` if this is a signed integer type like [`Type::I8`].
    #[must_use]
    pub const fn is_signed_integer(&self) -> bool {
        use Type::{I16, I32, I64, I8};
        matches!(self, I8 | I16 | I32 | I64)
    }

    /// Returns `true` if this is an unsigned integer type like [`Type::U8`].
    #[must_use]
    pub const fn is_unsigned_integer(&self) -> bool {
        use Type::{U16, U32, U64, U8};
        matches!(self, U8 | U16 | U32 | U64)
    }
}
