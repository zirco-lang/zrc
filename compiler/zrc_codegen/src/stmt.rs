//! Code generation for statements

use inkwell::{
    basic_block::BasicBlock,
    builder::Builder,
    context::Context,
    memory_buffer::MemoryBuffer,
    module::Module,
    passes::{PassManager, PassManagerBuilder},
    targets::{
        CodeModel, FileType, InitializationConfig, RelocMode, Target, TargetMachine, TargetTriple,
    },
    types::AnyType,
    values::{BasicValueEnum, FunctionValue},
    OptimizationLevel,
};
use zrc_typeck::tast::{
    expr::{Place, PlaceKind, TypedExpr, TypedExprKind},
    stmt::{ArgumentDeclaration, LetDeclaration, TypedDeclaration, TypedStmt},
    ty::Type,
};

use crate::{
    expr::cg_expr,
    ty::{create_fn, llvm_basic_type, llvm_type},
    CgScope,
};

/// Consists of the [`BasicBlock`]s to `br` to when encountering certain
/// instructions. It is passed to [`cg_block`] to allow it to properly handle
/// break and continue.
#[derive(PartialEq, Eq, Debug, Clone)]
struct LoopBreakaway<'ctx> {
    /// Points to the exit basic block.
    on_break: BasicBlock<'ctx>,
    /// For `for` loops, points to the latch. For `while` loops, points to the
    /// header.
    on_continue: BasicBlock<'ctx>,
}

/// Generates the `alloca`tion, `store` instruction, and adds a new identifier
/// to the [`CgScope`].
///
/// # Panics
/// Panics if an internal code generation error is encountered.
#[allow(clippy::trivially_copy_pass_by_ref, clippy::too_many_arguments)]
fn cg_let_declaration<'ctx, 'input, 'a>(
    ctx: &'ctx Context,
    target_machine: &TargetMachine,
    builder: &'a Builder<'ctx>,
    module: &'a Module<'ctx>,
    function: &'a FunctionValue<'ctx>,
    bb: &'a BasicBlock<'ctx>,
    scope: &'a mut CgScope<'input, 'ctx>,
    declarations: Vec<LetDeclaration<'input>>,
) -> BasicBlock<'ctx> {
    let mut bb = *bb;

    for let_declaration in declarations {
        let ptr = builder
            .build_alloca(
                llvm_basic_type(ctx, target_machine, &let_declaration.ty),
                &format!("let_{}", let_declaration.name),
            )
            .unwrap();

        scope.insert(let_declaration.name, ptr);

        if let Some(value) = let_declaration.value {
            bb = cg_expr(
                ctx,
                target_machine,
                builder,
                module,
                function,
                &bb,
                scope,
                TypedExpr(
                    let_declaration.ty.clone(),
                    TypedExprKind::Assignment(
                        Box::new(Place(
                            let_declaration.ty,
                            PlaceKind::Variable(let_declaration.name),
                        )),
                        Box::new(value),
                    ),
                ),
            )
            .1;
        }
    }

    bb
}

/// Process a vector of [`TypedStmt`]s (a block) and handle each statement.
///
/// # Panics
/// Panics if an internal code generation error is encountered.
#[allow(
    clippy::too_many_arguments,
    clippy::too_many_lines,
    clippy::needless_pass_by_value,
    clippy::trivially_copy_pass_by_ref
)]
fn cg_block<'ctx, 'input, 'a>(
    ctx: &'ctx Context,
    target_machine: &TargetMachine,
    builder: &'a Builder<'ctx>,
    module: &'a Module<'ctx>,
    function: &'a FunctionValue<'ctx>,
    bb: &'a BasicBlock<'ctx>,
    parent_scope: &'a CgScope<'input, 'ctx>,
    block: Vec<TypedStmt<'input>>,
    breakaway: &Option<LoopBreakaway<'ctx>>,
) -> Option<BasicBlock<'ctx>> {
    let mut scope = parent_scope.clone();

    block
        .into_iter()
        .try_fold(*bb, |bb, stmt| -> Option<BasicBlock> {
            match stmt {
                TypedStmt::ExprStmt(expr) => Some(
                    cg_expr(
                        ctx,
                        target_machine,
                        builder,
                        module,
                        function,
                        &bb,
                        &scope,
                        expr,
                    )
                    .1,
                ),

                TypedStmt::IfStmt(cond, then, then_else) => {
                    let then_else = then_else.unwrap_or(vec![]);

                    let (cond, _) = cg_expr(
                        ctx,
                        target_machine,
                        builder,
                        module,
                        function,
                        &bb,
                        &scope,
                        cond,
                    );

                    let then_bb = ctx.append_basic_block(*function, "then");
                    let then_else_bb = ctx.append_basic_block(*function, "then_else");

                    builder
                        .build_conditional_branch(cond.into_int_value(), then_bb, then_else_bb)
                        .unwrap();

                    builder.position_at_end(then_bb);
                    let maybe_then_bb = cg_block(
                        ctx,
                        target_machine,
                        builder,
                        module,
                        function,
                        &then_bb,
                        &scope,
                        then,
                        breakaway,
                    );

                    builder.position_at_end(then_else_bb);
                    let maybe_then_else_bb = cg_block(
                        ctx,
                        target_machine,
                        builder,
                        module,
                        function,
                        &then_else_bb,
                        &scope,
                        then_else,
                        breakaway,
                    );

                    match (maybe_then_bb, maybe_then_else_bb) {
                        (None, None) => None,
                        (Some(single_bb), None) | (None, Some(single_bb)) => {
                            let end = ctx.append_basic_block(*function, "end");

                            builder.position_at_end(single_bb);
                            builder.build_unconditional_branch(end).unwrap();

                            builder.position_at_end(end);
                            Some(end)
                        }
                        (Some(then_bb), Some(then_else_bb)) => {
                            let end = ctx.append_basic_block(*function, "end");

                            builder.position_at_end(then_bb);
                            builder.build_unconditional_branch(end).unwrap();

                            builder.position_at_end(then_else_bb);
                            builder.build_unconditional_branch(end).unwrap();

                            builder.position_at_end(end);
                            Some(end)
                        }
                    }
                }

                TypedStmt::BlockStmt(block) => cg_block(
                    ctx,
                    target_machine,
                    builder,
                    module,
                    function,
                    &bb,
                    &scope,
                    block,
                    breakaway,
                ),

                TypedStmt::ReturnStmt(Some(expr)) => {
                    let (expr, _) = cg_expr(
                        ctx,
                        target_machine,
                        builder,
                        module,
                        function,
                        &bb,
                        &scope,
                        expr,
                    );

                    builder.build_return(Some(&expr)).unwrap();

                    None
                }

                TypedStmt::ReturnStmt(None) => {
                    builder.build_return(None).unwrap();

                    None
                }

                TypedStmt::ContinueStmt => {
                    builder
                        .build_unconditional_branch(breakaway.as_ref().unwrap().on_continue)
                        .unwrap();

                    None
                }

                TypedStmt::BreakStmt => {
                    builder
                        .build_unconditional_branch(breakaway.as_ref().unwrap().on_break)
                        .unwrap();

                    None
                }

                TypedStmt::DeclarationList(declarations) => Some(cg_let_declaration(
                    ctx,
                    target_machine,
                    builder,
                    module,
                    function,
                    &bb,
                    &mut scope,
                    declarations,
                )),

                TypedStmt::ForStmt {
                    init,
                    cond,
                    post,
                    body,
                } => {
                    // For loops generate a somewhat more complicated CFG, with a few parts.
                    // The preheader, where `init` runs. Breaks to the header.
                    // The header, where `cond` is checked and breaks to either the exit or the
                    // body. The body, where most of the body runs. Breaks to
                    // the latch. `break` transfers to the exit by force and `continue` transfers to
                    // the latch by force. The latch, where `post` runs and
                    // breaks back to the header The exit, which is the basic
                    // block we return.

                    // loops lie in an implicit subscope
                    let mut scope = scope.clone();

                    // The block we are currently in will become the preheader. Generate the `init`
                    // code if there is any.
                    if let Some(init) = init {
                        cg_let_declaration(
                            ctx,
                            target_machine,
                            builder,
                            module,
                            function,
                            &bb,
                            &mut scope,
                            *init,
                        );
                    }

                    let header = ctx.append_basic_block(*function, "header");
                    let body_bb = ctx.append_basic_block(*function, "body");
                    let latch = ctx.append_basic_block(*function, "latch");
                    let exit = ctx.append_basic_block(*function, "exit");

                    // Branch to the header from the preheader.
                    builder.build_unconditional_branch(header).unwrap();

                    // Generate the header.
                    builder.position_at_end(header);
                    let header = cond.map_or_else(
                        || {
                            // If there is no condition, we always branch to the body.
                            builder.build_unconditional_branch(body_bb).unwrap();

                            header
                        },
                        |cond| {
                            let (cond, header) = cg_expr(
                                ctx,
                                target_machine,
                                builder,
                                module,
                                function,
                                &header,
                                &scope,
                                cond,
                            );

                            builder
                                .build_conditional_branch(cond.into_int_value(), body_bb, exit)
                                .unwrap();

                            header
                        },
                    );

                    // Generate the body.
                    builder.position_at_end(body_bb);
                    let body_bb = cg_block(
                        ctx,
                        target_machine,
                        builder,
                        module,
                        function,
                        &body_bb,
                        &scope,
                        body,
                        &Some(LoopBreakaway {
                            on_break: exit,
                            on_continue: latch,
                        }),
                    );

                    // The body breaks to latch
                    if body_bb.is_some() {
                        builder.build_unconditional_branch(latch).unwrap();
                    }

                    // Latch runs post and then breaks right back to the header.
                    builder.position_at_end(latch);
                    if let Some(post) = post {
                        cg_expr(
                            ctx,
                            target_machine,
                            builder,
                            module,
                            function,
                            &latch,
                            &scope,
                            post,
                        );
                    }

                    builder.build_unconditional_branch(header).unwrap();

                    builder.position_at_end(exit);

                    Some(exit)
                }

                TypedStmt::WhileStmt(cond, body) => {
                    // While loops are similar to for loops but much simpler.
                    // The preheader simply just breaks to the header.
                    // The header checks the condition and breaks to the exit or the body.
                    // The body simply breaks to the header.
                    // The exit is the continued code

                    // `break` => exit
                    // `continue` => header

                    let header = ctx.append_basic_block(*function, "header");

                    let body_bb = ctx.append_basic_block(*function, "body");

                    let exit = ctx.append_basic_block(*function, "exit");

                    builder.build_unconditional_branch(header).unwrap();

                    builder.position_at_end(header);

                    let (cond, _) = cg_expr(
                        ctx,
                        target_machine,
                        builder,
                        module,
                        function,
                        &header,
                        &scope,
                        cond,
                    );

                    builder
                        .build_conditional_branch(cond.into_int_value(), body_bb, exit)
                        .unwrap();

                    builder.position_at_end(body_bb);

                    let body_bb = cg_block(
                        ctx,
                        target_machine,
                        builder,
                        module,
                        function,
                        &body_bb,
                        &scope,
                        body,
                        &Some(LoopBreakaway {
                            on_break: exit,
                            on_continue: header,
                        }),
                    );

                    if body_bb.is_some() {
                        builder.build_unconditional_branch(header).unwrap();
                    }

                    builder.position_at_end(exit);

                    Some(exit)
                }
            }
        })
}

/// Initialize the LLVM [`FunctionValue`] for a given function prototype
pub fn cg_init_fn<'ctx>(
    ctx: &'ctx Context,
    module: &Module<'ctx>,
    target_machine: &TargetMachine,
    name: &str,
    ret: Option<Type>,
    args: &[&Type],
    is_variadic: bool,
) -> FunctionValue<'ctx> {
    let ret_type = llvm_type(ctx, target_machine, &ret.unwrap_or(Type::Void));
    let arg_types = args
        .iter()
        .map(|ty| llvm_basic_type(ctx, target_machine, ty).into())
        .collect::<Vec<_>>();

    let fn_type = create_fn(
        ret_type.as_any_type_enum(),
        arg_types.as_slice(),
        is_variadic,
    );
    let fn_val = module.add_function(name, fn_type, None);

    fn_val
}

/// Code generate and verify a program given a [`Context`] and return the final
/// global [`CgScope`], and [`Module`].
///
/// # Panics
/// Panics if code generation fails. This can be caused by an invalid TAST being
/// passed, so make sure to type check it so invariants are upheld.
#[must_use]
fn cg_program<'input, 'ctx>(
    ctx: &'ctx Context,
    target_machine: &TargetMachine,
    module_name: &str,
    program: Vec<TypedDeclaration<'input>>,
) -> (Module<'ctx>, CgScope<'input, 'ctx>) {
    let builder = ctx.create_builder();
    let module = ctx.create_module(module_name);

    let mut global_scope = CgScope::new();

    for declaration in program {
        match declaration {
            TypedDeclaration::FunctionDeclaration {
                name,
                parameters,
                return_type,
                body,
            } => {
                let fn_value = cg_init_fn(
                    ctx,
                    &module,
                    target_machine,
                    name,
                    return_type,
                    &parameters
                        .as_arguments()
                        .iter()
                        .map(|ArgumentDeclaration { ty, .. }| ty)
                        .collect::<Vec<_>>(),
                    parameters.is_variadic(),
                );
                global_scope.insert(name, fn_value.as_global_value().as_pointer_value());
                // must come after the insert call so that recursion is valid
                let mut fn_scope = global_scope.clone();

                if let Some(body) = body {
                    let entry = ctx.append_basic_block(fn_value, "entry");
                    builder.position_at_end(entry);

                    for (n, ArgumentDeclaration { name, ty }) in
                        parameters.into_arguments().into_iter().enumerate()
                    {
                        if entry.get_first_instruction().is_some() {
                            builder.position_before(&entry.get_first_instruction().unwrap());
                        } else {
                            builder.position_at_end(entry);
                        }

                        let alloc = builder
                            .build_alloca(
                                llvm_basic_type(ctx, target_machine, &ty),
                                &format!("arg_{name}"),
                            )
                            .unwrap();

                        builder.position_at_end(entry);

                        builder
                            .build_store::<BasicValueEnum>(
                                alloc,
                                fn_value
                                    .get_nth_param(
                                        n.try_into()
                                            .expect("over u32::MAX parameters in a function? HOW?"),
                                    )
                                    .unwrap(),
                            )
                            .unwrap();

                        fn_scope.insert(name, alloc);
                    }

                    cg_block(
                        ctx,
                        target_machine,
                        &builder,
                        &module,
                        &fn_value,
                        &entry,
                        &fn_scope,
                        body,
                        &None,
                    );
                }
            }
        }
    }

    match module.verify() {
        Ok(()) => {}

        Err(error_as_llvm_string) => {
            panic!(
                "code generation failure:\n{}\nGenerated IR:\n{}",
                error_as_llvm_string.to_string(),
                module.print_to_string().to_string()
            );
        }
    }

    module.verify().expect("Generated invalid LLVM IR");

    (module, global_scope)
}

/// Run optimizations on the given program.
fn optimize_module(module: &Module<'_>, optimization_level: OptimizationLevel) {
    let pmb = PassManagerBuilder::create();
    pmb.set_optimization_level(optimization_level);
    let pm = PassManager::<Module>::create(());
    pmb.populate_module_pass_manager(&pm);
    pm.run_on(module);
}

/// Code generate a LLVM program to a string.
///
/// # Panics
/// Panics on internal code generation failure.
#[must_use]
pub fn cg_program_to_string(
    module_name: &str,
    program: Vec<TypedDeclaration>,
    optimization_level: OptimizationLevel,
    triple: &TargetTriple,
    cpu: &str,
) -> String {
    let ctx = Context::create();

    Target::initialize_all(&InitializationConfig::default());
    let target = Target::from_triple(triple).unwrap();

    let target_machine = target
        .create_target_machine(
            triple,
            cpu,
            "",
            // FIXME: Does this potentially run the optimizer twice? That may be inefficient.
            optimization_level,
            RelocMode::PIC,
            CodeModel::Default,
        )
        .unwrap();

    let (module, _global_scope) = cg_program(&ctx, &target_machine, module_name, program);
    optimize_module(&module, optimization_level);

    module.print_to_string().to_string()
}

/// Code generate a LLVM program to a [`MemoryBuffer`] based on the given
/// [`FileType`].
///
/// # Panics
/// Panics on internal code generation failure.
#[must_use]
pub fn cg_program_to_buffer(
    module_name: &str,
    program: Vec<TypedDeclaration>,
    file_type: FileType,
    optimization_level: OptimizationLevel,
    triple: &TargetTriple,
    cpu: &str,
) -> MemoryBuffer {
    let ctx = Context::create();

    Target::initialize_all(&InitializationConfig::default());
    let target = Target::from_triple(triple).unwrap();

    let target_machine = target
        .create_target_machine(
            triple,
            cpu,
            "",
            // FIXME: Does this potentially run the optimizer twice? That may be inefficient.
            optimization_level,
            RelocMode::PIC,
            CodeModel::Default,
        )
        .unwrap();

    let (module, _global_scope) = cg_program(&ctx, &target_machine, module_name, program);
    optimize_module(&module, optimization_level);

    target_machine
        .write_to_memory_buffer(&module, file_type)
        .unwrap()
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn let_declarations_are_properly_generated() {
        cg_snapshot_test!(indoc! {"
            fn test() {
                // TEST: should allocate twice and assign to one.
                let a: i32;
                let b: i32 = 7;

                return;
            }
        "});
    }

    mod cg_block {
        use super::*;

        mod conditionals {
            use super::*;

            #[test]
            fn if_statements_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;
                    fn nop();

                    fn test() {
                        // TEST: properly produces a conditional break over the call result and
                        // both code paths join at the end
                        if (get_bool()) nop();

                        // TEST: code generation properly continues in the last block
                        nop();
                        return;
                    }
                "});
            }

            #[test]
            fn if_else_statements_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;
                    fn nop();

                    fn test() {
                        // TEST: properly produces a conditional break over the call result and
                        // both code baths call nop().
                        if (get_bool()) nop();
                        else {
                            nop();
                            // TEST: this path diverges
                            return;
                        }

                        // TEST: code generation properly continues from the if_true block
                        nop();
                        return;
                    }
                "});
            }

            #[test]
            fn if_else_statements_where_both_blocks_terminate_do_not_continue_generating() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: properly produces a conditional break over the call result
                        // and both code paths return (diverge). there should be no %end bb.
                        if (get_bool()) return;
                        else return;
                    }
                "});
            }
        }

        mod loops {
            use super::*;

            #[test]
            fn while_loops_along_with_break_and_continue_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: the proper while loop structure is created
                        while (get_bool()) {
                            // TEST: break jumps to the `end` block
                            if (get_bool()) break;
                            else {
                                // TEST: continue jumps to the header block
                                if (get_bool()) continue;
                                // TEST: otherwise, we proceed
                                else {}
                            }

                            // TEST: the loop jumps back to the header block
                        }

                        // TEST: ...and code generation properly continues.
                        return;
                    }
                "});
            }

            #[test]
            fn for_loops_along_with_break_and_continue_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_int() -> i32;

                    fn test() {
                        // TEST: the proper while loop structure is created
                        for (let i = 0; i < get_int(); i += 1) {
                            // TEST: break jumps to the `end` block
                            if (i > get_int()) break;
                            else {
                                // TEST: continue jumps to the latch block
                                if (i < get_int()) continue;
                                else {}
                            }

                            // TEST: the loop jumps to the latch block which jumps back to the
                            // header
                        }

                        return;
                    }
                "});
            }
        }
    }
}
