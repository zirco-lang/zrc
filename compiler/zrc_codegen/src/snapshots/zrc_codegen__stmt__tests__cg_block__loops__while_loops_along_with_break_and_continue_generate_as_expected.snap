---
source: compiler/zrc_codegen/src/stmt.rs
description: "fn get_bool() -> bool;\n\nfn test() {\n    // TEST: the proper while loop structure is created\n    while (get_bool()) {\n        // TEST: break jumps to the `end` block\n        if (get_bool()) break;\n        else {\n            // TEST: continue jumps to the header block\n            if (get_bool()) continue;\n            // TEST: otherwise, we proceed\n            else {}\n        }\n\n        // TEST: the loop jumps back to the header block\n    }\n\n    // TEST: ...and code generation properly continues.\n    return;\n}\n"
expression: resulting_ir
---
; ModuleID = 'test'
source_filename = "test"

declare i1 @get_bool()

define void @test() {
entry:
  br label %header

header:                                           ; preds = %end5, %then3, %entry
  %call = call i1 @get_bool()
  br i1 %call, label %body, label %exit

body:                                             ; preds = %header
  %call1 = call i1 @get_bool()
  br i1 %call1, label %then, label %then_else

exit:                                             ; preds = %then, %header
  ret void

then:                                             ; preds = %body
  br label %exit

then_else:                                        ; preds = %body
  %call2 = call i1 @get_bool()
  br i1 %call2, label %then3, label %then_else4

then3:                                            ; preds = %then_else
  br label %header

then_else4:                                       ; preds = %then_else
  br label %end

end:                                              ; preds = %then_else4
  br label %end5

end5:                                             ; preds = %end
  br label %header
}

